matchbox_signaling/src/topologies/client_server.rs

```diff
@@ -0,0 +1,296 @@
+use crate::{
+    signaling_server::{
+        error::{ClientRequestError, SignalingError},
+        handlers::WsStateMeta,
+        SignalingState,
+    },
+    topologies::{
+        common_logic::{parse_request, spawn_sender_task},
+        SignalingTopology,
+    },
+    Callback, SignalingCallbacks, SignalingServerBuilder,
+};
+use async_trait::async_trait;
+use axum::extract::ws::Message;
+use futures::StreamExt;
+use matchbox_protocol::{JsonPeerEvent, PeerId, PeerRequest};
+use std::collections::HashMap;
+use tracing::{error, info, warn};
+
+use super::common_logic::{try_send, SignalingChannel, StateObj};
+
+#[derive(Debug, Default)]
+pub struct ClientServer;
+
+impl SignalingServerBuilder<ClientServer, ClientServerCallbacks, ClientServerState> {
+    /// Set a callback triggered on all client websocket connections.
+    pub fn on_client_connected<F>(mut self, callback: F) -> Self
+    where
+        F: Fn(PeerId) + 'static,
+    {
+        self.callbacks.on_client_connected = Callback::from(callback);
+        self
+    }
+
+    /// Set a callback triggered on all client websocket disconnections.
+    pub fn on_client_disconnected<F>(mut self, callback: F) -> Self
+    where
+        F: Fn(PeerId) + 'static,
+    {
+        self.callbacks.on_client_disconnected = Callback::from(callback);
+        self
+    }
+
+    /// Set a callback triggered on host websocket connection.
+    pub fn on_host_connected<F>(mut self, callback: F) -> Self
+    where
+        F: Fn(PeerId) + 'static,
+    {
+        self.callbacks.on_host_connected = Callback::from(callback);
+        self
+    }
+
+    /// Set a callback triggered on host websocket disconnection.
+    pub fn on_host_disconnected<F>(mut self, callback: F) -> Self
+    where
+        F: Fn(PeerId) + 'static,
+    {
+        self.callbacks.on_host_disconnected = Callback::from(callback);
+        self
+    }
+}
+
+#[async_trait]
+impl SignalingTopology<ClientServerCallbacks, ClientServerState> for ClientServer {
+    async fn state_machine(upgrade: WsStateMeta<ClientServerCallbacks, ClientServerState>) {
+        let WsStateMeta {
+            ws,
+            mut state,
+            callbacks,
+        } = upgrade;
+
+        let (ws_sender, mut ws_receiver) = ws.split();
+        let sender = spawn_sender_task(ws_sender);
+
+        // Generate a UUID for the user
+        let peer_uuid = uuid::Uuid::new_v4().into();
```

Comment:
Should this be offered as a callback? Might be nice if somebody wanted to add the ability to resume sessions or use meaningful uuids

Reply:
Yea, that's fair. Will do.